pkg/shardddl/pessimism/operation.go

```diff
@@ -0,0 +1,226 @@
+	DDLs   []string `json:"ddls"`   // DDL statements
+	Exec   bool     `json:"exec"`   // execute or skip the DDL statements
+	Done   bool     `json:"done"`   // whether the `Exec` operation has done
+}
+
+// NewOperation creates a new Operation instance.
+func NewOperation(ID, task, source string, DDLs []string, exec, done bool) Operation {
+	return Operation{
+		ID:     ID,
+		Task:   task,
+		Source: source,
+		DDLs:   DDLs,
+		Exec:   exec,
+		Done:   done,
+	}
+}
+
+// String implements Stringer interface.
+func (o Operation) String() string {
+	s, _ := o.toJSON()
+	return s
+}
+
+// toJSON returns the string of JSON represent.
+func (o Operation) toJSON() (string, error) {
+	data, err := json.Marshal(o)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+// operationFromJSON constructs Operation from its JSON represent.
+func operationFromJSON(s string) (o Operation, err error) {
+	err = json.Unmarshal([]byte(s), &o)
+	return
+}
+
+// PutOperations puts the shard DDL operations into etcd.
+// if `skipDone` is `true`:
+//   - PUT: all of kvs ("not exist" or "the `done` field is not `true`")
+//   - skip PUT: any of kvs ("exist" and "the `done` field is `true`")
+// NOTE:
+//   `clientv3.Value` has a strange behavior for *not-exist* kv,
+//   see https://github.com/etcd-io/etcd/issues/10566.
+//   In addition, etcd compare has no `OR` operator now,
+//   see https://github.com/etcd-io/etcd/issues/10571.
+//   So, it's hard to do our `skipDone` logic in one txn.
+//   We break the logic into two txn, but this may lead to problem when PUT operations concurrently.
+// This function should often be called by DM-master.
+func PutOperations(cli *clientv3.Client, skipDone bool, ops ...Operation) (rev int64, putted bool, err error) {
+	cmpsNotExist := make([]clientv3.Cmp, 0, len(ops))
+	cmpsNotDone := make([]clientv3.Cmp, 0, len(ops))
+	opsPut := make([]clientv3.Op, 0, len(ops))
+	for _, op := range ops {
+		value, err2 := op.toJSON()
+		if err2 != nil {
+			return 0, false, err2
+		}
+
+		key := common.ShardDDLPessimismOperationKeyAdapter.Encode(op.Task, op.Source)
+		opsPut = append(opsPut, clientv3.OpPut(key, value))
+
+		if skipDone {
+			opDone := op
+			opDone.Done = true // set `done` to `true`.
+			valueDone, err3 := opDone.toJSON()
+			if err3 != nil {
+				return 0, false, err3
+			}
+			cmpsNotExist = append(cmpsNotExist, clientv3util.KeyMissing(key))
+			cmpsNotDone = append(cmpsNotDone, clientv3.Compare(clientv3.Value(key), "!=", valueDone))
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(cli.Ctx(), etcdutil.DefaultRequestTimeout)
+	defer cancel()
+
+	// txn 1: try to PUT if all of kvs "not exist".
+	resp, err := cli.Txn(ctx).If(cmpsNotExist...).Then(opsPut...).Commit()
+	if err != nil {
+		return 0, false, err
+	} else if resp.Succeeded {
+		return resp.Header.Revision, resp.Succeeded, nil
+	}
+
+	// txn 2: try to PUT if all of kvs "the `done` field is not `true`.
```

Comment:
do you mean `txn 2: try to PUT if any of kvs "the `done` field is not `true``?

Reply:
No, it's `all` (`If(cmpsNotDone...)`), if `any` of kvs ``` "the `done` field is not `true`" ```, it should skip the PUT.